### PR TITLE
feat(events-targets): add role support for sqs target

### PR DESCRIPTION
**Pull request by seyeong**
----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [x] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by Naumel** [comment]
Hello,

Thank you for taking time and submitting changes for review!

Please be advised that some of the integration tests failed. You can see the full output [here](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e2a24bfe-cca2-49d2-8dd9-da2bf3dd05d0%2Fbuild.log).

Cheers!

**Comment by seyeong** [comment]
Hey,

I am having a trouble running `yarn integ`. Can you kindly run it and fix any issues for me?

**Comment by seyeong** [comment]
I stand corrected. It's failing in where I did not make any change.

```
@aws-cdk-containers/ecs-service-extensions: Error: integ-runner exited with error code 1
```

Integ tests for `aws-events-targets` are passing.


```
@aws-cdk/aws-events-targets:   UNCHANGED  batch/integ.job-definition-events 2.539s
@aws-cdk/aws-events-targets:   UNCHANGED  event-bus/integ.event-bus 2.542s
@aws-cdk/aws-events-targets:   UNCHANGED  codepipeline/integ.pipeline-event-target 2.637s
@aws-cdk/aws-events-targets:   UNCHANGED  lambda/integ.events 2.645s
@aws-cdk/aws-events-targets:   UNCHANGED  aws-api/integ.aws-api 2.649s
@aws-cdk/aws-events-targets:   UNCHANGED  kinesis-firehose/integ.kinesis-firehose-stream 2.715s
@aws-cdk/aws-events-targets:   UNCHANGED  kinesis/integ.kinesis-stream 2.643s
@aws-cdk/aws-events-targets:   UNCHANGED  sqs/integ.sqs-event-rule-target 2.529s
@aws-cdk/aws-events-targets:   UNCHANGED  sns/integ.sns-event-rule-target 2.583s
@aws-cdk/aws-events-targets:   UNCHANGED  codebuild/integ.project-events 2.912s
@aws-cdk/aws-events-targets: Snapshot Results: 
@aws-cdk/aws-events-targets: Tests:    13 passed, 13 total
@aws-cdk/aws-events-targets: Tests successful. Total time (14.7s) | /codebuild/output/src016164706/src/github.com/aws/aws-cdk/node_modules/jest/bin/jest.js (11.3s) | integ-runner (3.4s)
@aws-cdk/aws-events-targets: Done in 48.57s.
```

File: packages/@aws-cdk/aws-events-targets/lib/sqs.ts [hunk 1]
```diff
@@ -52,6 +59,10 @@ export class SqsQueue implements events.IRuleTarget {
    * @see https://docs.aws.amazon.com/eventbridge/latest/userguide/resource-based-policies-eventbridge.html#sqs-permissions
    */
   public bind(rule: events.IRule, _id?: string): events.RuleTargetConfig {
+    const role = this.props.role ?? singletonEventRole(rule);
```
Comment (otaviomacedo) [inline packages/@aws-cdk/aws-events-targets/lib/sqs.ts:62]: Why do we need to create a role when none has been provided? To ensure that we keep the current behavior, we should leave it undefined.

**Review by otaviomacedo — CHANGES_REQUESTED** [review]

**Comment by TheRealAmazonKendra** [comment]
> I stand corrected. It's failing in where I did not make any change.
> 
> ```
> @aws-cdk-containers/ecs-service-extensions: Error: integ-runner exited with error code 1
> ```
> 
> Integ tests for `aws-events-targets` are passing.
> 
> ```
> @aws-cdk/aws-events-targets:   UNCHANGED  batch/integ.job-definition-events 2.539s
> @aws-cdk/aws-events-targets:   UNCHANGED  event-bus/integ.event-bus 2.542s
> @aws-cdk/aws-events-targets:   UNCHANGED  codepipeline/integ.pipeline-event-target 2.637s
> @aws-cdk/aws-events-targets:   UNCHANGED  lambda/integ.events 2.645s
> @aws-cdk/aws-events-targets:   UNCHANGED  aws-api/integ.aws-api 2.649s
> @aws-cdk/aws-events-targets:   UNCHANGED  kinesis-firehose/integ.kinesis-firehose-stream 2.715s
> @aws-cdk/aws-events-targets:   UNCHANGED  kinesis/integ.kinesis-stream 2.643s
> @aws-cdk/aws-events-targets:   UNCHANGED  sqs/integ.sqs-event-rule-target 2.529s
> @aws-cdk/aws-events-targets:   UNCHANGED  sns/integ.sns-event-rule-target 2.583s
> @aws-cdk/aws-events-targets:   UNCHANGED  codebuild/integ.project-events 2.912s
> @aws-cdk/aws-events-targets: Snapshot Results: 
> @aws-cdk/aws-events-targets: Tests:    13 passed, 13 total
> @aws-cdk/aws-events-targets: Tests successful. Total time (14.7s) | /codebuild/output/src016164706/src/github.com/aws/aws-cdk/node_modules/jest/bin/jest.js (11.3s) | integ-runner (3.4s)
> @aws-cdk/aws-events-targets: Done in 48.57s.
> ```

These tests are using the `events-target` construct in their tests, so the changes here are causing that failure. This indicates that this change, as is, could cause breaking changes for current customers. @otaviomacedo's comment above addresses where that's being caused.

File: packages/@aws-cdk/aws-events-targets/lib/sqs.ts (hunk 1 above)
Comment (seyeong) [reply packages/@aws-cdk/aws-events-targets/lib/sqs.ts]: You're right. I have addressed your comment.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Please make sure that your PR body describes the problem the PR is solving, and the design approach and alternatives considered. Explain why the PR solves the problem. A link to an issue is helpful, but does not replace an explanation of your thought process (See [Contributing Guide, Pull Requests](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request)).

Additionally, this change needs an integration test before we can provide a meaningful review.

File: packages/@aws-cdk/aws-events-targets/lib/sqs.ts [hunk 2]
```diff
@@ -52,6 +59,12 @@ export class SqsQueue implements events.IRuleTarget {
    * @see https://docs.aws.amazon.com/eventbridge/latest/userguide/resource-based-policies-eventbridge.html#sqs-permissions
    */
   public bind(rule: events.IRule, _id?: string): events.RuleTargetConfig {
+    const role = this.props.role;
+
+    if (role !== undefined) {
+      role.addToPrincipalPolicy(this.sendMessageStatement());
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events-targets/lib/sqs.ts:65]: Why not use `this.queue.grantSendMessages(role)` ?

Comment (rix0rrr) [reply packages/@aws-cdk/aws-events-targets/lib/sqs.ts:65]: The `this.queue.grantSendMessages()` call a couple lines down should probably be repurposed for this, to look like:

```ts
this.queue.grantSendMessages(this.props.role ?? new iam.ServicePrincipal('events.amazonaws.com', principalOpts));
```

File: packages/@aws-cdk/aws-events-targets/README.md [hunk 3]
```diff
@@ -345,3 +342,25 @@ rule.addTarget(new targets.EventBus(
...
+Use the `SQS` target to route event to a SQS queue.
+
+The code snippet below creates the scheduled event rule that route events to a SQS queue.
+
+```ts
+const queue = new sqs.Queue(this, 'Queue');
+
+const role = new iam.Role(this, 'Role', {
+  assumedBy: new iam.ServicePrincipal('events.amazonaws.com'),
+});
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events-targets/README.md:357]: I'd rather not have this in the example if it's not necessary.

File: packages/@aws-cdk/aws-events-targets/lib/sqs.ts [hunk 4]
```diff
@@ -25,6 +25,13 @@ export interface SqsQueueProps extends TargetBaseProps {
    * @default the entire EventBridge event
    */
   readonly message?: events.RuleTargetInput;
+
+  /**
+   * Role to be used to publish the event
+   *
+   * @default - no role assigned to the target
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events-targets/lib/sqs.ts:32]: This sounds ominous. Any downsides to it?

**Review by rix0rrr — CHANGES_REQUESTED** [review]

**Pull request closed without merging**